pkg/agent/monitortool/latency_store.go

```diff
@@ -0,0 +1,271 @@
+
+	// Whether the agent is running in networkPolicyOnly mode
+	isNetworkPolicyOnly bool
+	// The map of Node IP to latency entry, it will be changed by latency monitor
+	nodeIPLatencyMap map[string]*NodeIPLatencyEntry
+	// The map of Node name to Node IP, it will be changed by Node watcher
+	// If the agent is running in networkPolicyOnly mode, the value will be the transport IP of the Node
+	// If the agent is running in normal mode, the value will be the gateway IP of the Node
+	nodeTargetIPsMap map[string][]net.IP
+}
+
+// NodeIPLatencyEntry is the entry of the latency map.
+type NodeIPLatencyEntry struct {
+	// The timestamp of the last send packet
+	LastSendTime time.Time
+	// The timestamp of the last receive packet
+	LastRecvTime time.Time
+	// The last valid rtt of the connection
+	LastMeasuredRTT time.Duration
+}
+
+// NewLatencyStore creates a new LatencyStore.
+func NewLatencyStore(isNetworkPolicyOnly bool) *LatencyStore {
+	store := &LatencyStore{
+		nodeIPLatencyMap:    make(map[string]*NodeIPLatencyEntry),
+		nodeTargetIPsMap:    make(map[string][]net.IP),
+		isNetworkPolicyOnly: isNetworkPolicyOnly,
+	}
+
+	return store
+}
+
+// getNodeIPLatencyEntry returns the NodeIPLatencyEntry for the given Node IP
+// Now it used for testing purpose and it is not exported, so we just return a pointer
+func (l *LatencyStore) getNodeIPLatencyEntry(key string) (*NodeIPLatencyEntry, bool) {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+	entry, ok := l.nodeIPLatencyMap[key]
+	return entry, ok
+}
+
+// DeleteNodeIPLatencyEntry deletes the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) DeleteNodeIPLatencyEntry(key string) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	delete(l.nodeIPLatencyMap, key)
+}
+
+// SetNodeIPLatencyEntry sets the NodeIPLatencyEntry for the given Node IP
+func (l *LatencyStore) SetNodeIPLatencyEntry(key string, mutator func(entry *NodeIPLatencyEntry)) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	entry, ok := l.nodeIPLatencyMap[key]
+	if !ok {
+		// Init the connection entry
+		entry = &NodeIPLatencyEntry{}
+		l.nodeIPLatencyMap[key] = entry
+	}
+
+	// Update the connection entry
+	mutator(entry)
+}
+
+// ListLatencies returns the map of Node IP to latency entry
+func (l *LatencyStore) ListLatencies() map[string]*NodeIPLatencyEntry {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	return l.nodeIPLatencyMap
+}
+
+// addNode adds a Node to the latency store
+func (l *LatencyStore) addNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	l.updateNodeMap(node)
+}
+
+// deleteNode deletes a Node from the latency store
+func (l *LatencyStore) deleteNode(node *corev1.Node) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	// Delete from the node IP map
+	delete(l.nodeTargetIPsMap, node.Name)
+
+	// Delete from the NodeIPLatencyEntry map
+	getNodeIPs, err := l.getNodeIPs(node)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get IPs for Node", "nodeName", node.Name)
+		return
+	}
```

Comment:
It would be much better not to rely on this for deletion. Ideally you should do your cleanup simply based on the Node name and your internal state.

Reply:
Not addressed, we should not need to call `l.getNodeIPs(node)` in the deletion handler. We should do a lookup internally based on the Node name, and figure out what we need to clean up.

Reply:
OK, I use `nodeTargetIPsMap` to retrieve target IPs and remove them.